python/semantic_kernel/utils/tracing/decorators.py

```diff
@@ -0,0 +1,233 @@
+
+    def inner_trace_chat_completion(completion_func: Callable) -> Callable:
+        @functools.wraps(completion_func)
+        async def wrapper_decorator(*args: Any, **kwargs: Any) -> list[ChatMessageContent]:
+            chat_history: ChatHistory = kwargs["chat_history"]
+            settings: PromptExecutionSettings = kwargs["settings"]
+
+            if hasattr(settings, "ai_model_id") and settings.ai_model_id:
+                model_name = settings.ai_model_id
+            elif hasattr(args[0], "ai_model_id") and args[0].ai_model_id:
+                model_name = args[0].ai_model_id
+            else:
+                model_name = "unknown"
+
+            span = _start_completion_activity(model_name, model_provider, chat_history, settings)
+
+            try:
+                completions: list[ChatMessageContent] = await completion_func(*args, **kwargs)
+            except Exception as exception:
+                if span:
+                    _set_completion_error(span, exception)
+                    span.end()
+                raise
+
+            if span:
+                with trace.use_span(span, end_on_exit=True):
+                    if completions:
+                        first_completion = completions[0]
+                        response_id = first_completion.metadata.get("id", None)
+                        if not response_id:
+                            response_id = (
+                                first_completion.inner_content.get("id", None)
+                                if first_completion.inner_content
+                                else None
+                            )
+                        usage = first_completion.metadata.get("usage", None)
+                        prompt_tokens = usage.prompt_tokens if hasattr(usage, "prompt_tokens") else None
+                        completion_tokens = usage.completion_tokens if hasattr(usage, "completion_tokens") else None
+                    _set_completion_response(
+                        span, completions, response_id or "unknown", prompt_tokens, completion_tokens
+                    )
+
+            return completions
+
+        return wrapper_decorator
+
+    return inner_trace_chat_completion
+
+
+def _start_completion_activity(
+    model_name: str, model_provider: str, chat_history: ChatHistory, execution_settings: PromptExecutionSettings | None
+) -> Span | None:
+    """Start a text or chat completion activity for a given model."""
+    if not are_model_diagnostics_enabled():
+        return None
+
+    operation_name: str = "chat.completions"
+
+    span = tracer.start_span(f"{operation_name} {model_name}")
+
+    # Set attributes on the span
+    span.set_attributes(
+        {
+            OPERATION: operation_name,
+            SYSTEM: model_provider,
+            MODEL: model_name,
+        }
+    )
+
+    if execution_settings:
+        attribute = execution_settings.extension_data.get("max_tokens")
```

Comment:
how will we deal with some models having different names for this and other params?

Reply:
This is why this code was originally not a decorator.

What I see going forward is functions that extract params by trying a list of potential names.   Not awesome. I'm open to suggestions here!

Reply:
could you add a TODO for this so that whoever picks up this to apply it to other models can check and figure out something?

Reply:
I'm now thinking maybe we need to add a method to the PromptExecutionSettings class (overridden by each type), that maps itself from self to the different OTel keys...

Reply:
Added a TODO